src/relations/mysql_tls.py

```diff
@@ -0,0 +1,237 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of the tls certificates relation."""
+
+
+import base64
+import logging
+import os
+import re
+import socket
+from string import Template
+from typing import List, Optional, Tuple
+
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateAvailableEvent,
+    CertificateExpiringEvent,
+    TLSCertificatesRequiresV1,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import ActionEvent, CharmBase
+from ops.framework import Object
+from ops.model import MaintenanceStatus
+
+from constants import TLS_RELATION, TLS_SSL_CA_FILE, TLS_SSL_CERT_FILE, TLS_SSL_KEY_FILE
+from mysqlsh_helpers import (
+    MYSQL_DATA_DIR,
+    MYSQLD_CONFIG_DIRECTORY,
+    write_content_to_file,
+)
+
+logger = logging.getLogger(__name__)
+
+SCOPE = "unit"
+
+
+class MySQLTLS(Object):
+    """MySQL TLS Provider class."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "certificates")
+        self.charm = charm
+
+        self.certs = TLSCertificatesRequiresV1(self.charm, TLS_RELATION)
+
+        self.framework.observe(
+            self.charm.on.set_tls_private_key_action,
+            self._on_set_tls_private_key,
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_joined, self._on_tls_relation_joined
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_broken, self._on_tls_relation_broken
+        )
+
+        self.framework.observe(self.certs.on.certificate_available, self._on_certificate_available)
+        self.framework.observe(self.certs.on.certificate_expiring, self._on_certificate_expiring)
+
+    # =======================
+    #  Event Handlers
+    # =======================
+    def _on_set_tls_private_key(self, event: ActionEvent) -> None:
+        """Action for setting a TLS private key."""
+        self._request_certificate(event.params.get("internal-key", None))
```

Comment:
is it worth anything to set the `internal-key` into a secret?

Reply:
Yes, and it's done on the `self._request_certificate`